File: verifier/src/jsontext.ts
// SPDX-License-Identifier: Apache-2.0
//
// Two properties of a receipts file cannot be checked after parsing,
// because parsing destroys the evidence. Both are checked here, on the
// text as received, and the Go module checks the same two in
// receipts/jsontext.go.
//
// Duplicate member names: every parser keeps one value and discards the
// other, and they do not all keep the same one. JSON.parse silently
// takes the last. A document with a duplicate therefore means different
// things to different verifiers while the signature covers only one of
// those meanings, so it is refused rather than resolved.
//
// Lone surrogates: a \uD800-\uDFFF escape with no partner is not a
// character. JavaScript preserves it and Go replaces it with U+FFFD, so
// the two canonicalize different strings and reach different digests for
// the same file. RFC 8785 requires rejection.

/** Returns the first problem with the raw text, or null. */
export function jsonTextProblem(text: string): string | null {
  return loneSurrogateProblem(text) ?? numberSpellingProblem(text) ?? duplicateMemberProblem(text);
}

/**
 * Every number in a receipt is an integer (SPEC section 4), and the
 * spelling is part of the wire form: 1, never 1.0 or 1e0. JSON.parse
 * turns all three into the same value, so the spelling has to be checked
 * on the text. Go's typed decoder refuses the other two, and without
 * this the browser would accept a file the library rejects.
 */
function numberSpellingProblem(text: string): string | null {
  let inString = false;
  for (let i = 0; i < text.length; i++) {
    const c = text[i];
    if (inString) {
      if (c === "\\") i++;
      else if (c === '"') inString = false;
      continue;
    }
    if (c === '"') {
      inString = true;
      continue;
    }
    if (c !== "-" && (c < "0" || c > "9")) continue;
    // A number token: read to its end and judge the whole of it.
    let j = i;
    if (text[j] === "-") j++;
    const start = j;
    while (j < text.length && /[0-9.eE+-]/.test(text[j])) j++;
    const token = text.slice(start, j);
    if (token.includes(".") || token.includes("e") || token.includes("E")) {
      return `${text.slice(i, j)} is written with a fractional part or an exponent; this format carries whole numbers only`;
    }
    i = j - 1;
  }
  return null;
}

function loneSurrogateProblem(text: string): string | null {
  for (let i = 0; i + 5 < text.length; i++) {
    if (text[i] !== "\\" || text[i + 1] !== "u") continue;
    let backslashes = 0;
    for (let j = i - 1; j >= 0 && text[j] === "\\"; j--) backslashes++;
    if (backslashes % 2 === 1) continue; // the backslash is itself escaped

    const code = hex4(text.slice(i + 2, i + 6));
    if (code === null) continue; // malformed escapes are the parser's business
    const isHigh = code >= 0xd800 && code <= 0xdbff;
    const isLow = code >= 0xdc00 && code <= 0xdfff;
    if (!isHigh && !isLow) continue;

    if (isHigh && text[i + 6] === "\\" && text[i + 7] === "u") {
      const next = hex4(text.slice(i + 8, i + 12));
      if (next !== null && next >= 0xdc00 && next <= 0xdfff) {
        i += 11; // a well-formed pair
        continue;
      }
    }
    return `the file contains a lone surrogate escape (\\u${code.toString(16).toUpperCase().padStart(4, "0")}), which is not a character`;
  }
  return null;
}

function hex4(s: string): number | null {
  if (s.length !== 4 || !/^[0-9a-fA-F]{4}$/.test(s)) return null;
  return parseInt(s, 16);
}

/**
 * Walks the text and reports any object that names the same member
 * twice. This is a scanner rather than a parse: JSON.parse has already
 * thrown the duplicate away by the time a reviver sees the object.
 */
function duplicateMemberProblem(text: string): string | null {
  // Per open object: the names seen so far. null marks an open array.
  const stack: (Set<string> | null)[] = [];
  const path: string[] = [];
  let expectKey = false;
  let i = 0;

  const readString = (): string | null => {
    // text[i] is the opening quote. The raw span is collected and then
    // decoded, because two spellings of one name are one member: Go
    // compares decoded names, so "schema" and "\u0073chema" are a
    // duplicate there and would slip past a raw comparison here.
    const start = i;
    i++;
    while (i < text.length) {
      const c = text[i];
      if (c === "\\") {
        i += 2;
        continue;
      }
      if (c === '"') {
        i++;
        const raw = text.slice(start, i);
        try {
          return JSON.parse(raw) as string;
        } catch {
          return raw; // malformed: the parser will report it properly
        }
      }
      i++;
    }
    return null; // unterminated: the parser will report it
  };

  while (i < text.length) {
    const c = text[i];
    if (c === '"') {
      const name = readString();
      if (name === null) return null;
      const top = stack[stack.length - 1];
      if (expectKey && top) {
        if (top.has(name)) {
          const where = path.length > 0 ? path.join(".") : "the top level";
          return `"${name}" appears twice in ${where}; a duplicate member means different things to different parsers`;
        }
        top.add(name);
        path.push(name);
        expectKey = false;
      }
      continue;
    }
    if (c === "{") {
      stack.push(new Set());
      expectKey = true;
    } else if (c === "[") {
      stack.push(null);
      expectKey = false;
    } else if (c === "}" || c === "]") {
      stack.pop();
      path.pop();
      expectKey = stack.length > 0 && stack[stack.length - 1] !== null;
    } else if (c === ",") {
      const top = stack[stack.length - 1];
      if (top) {
        expectKey = true;
        path.pop();
      }
    }
    i++;
  }
  return null;
}

File: receipts/jsontext.go
// SPDX-License-Identifier: Apache-2.0

package receipts

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"strings"
	"unicode/utf16"
	"unicode/utf8"
)

// Two properties of a receipts file cannot be checked after parsing,
// because parsing destroys the evidence. Both are checked here, on the
// bytes as received.
//
// Duplicate member names: every JSON parser keeps one of the values and
// discards the other, and they do not all keep the same one. A document
// with a duplicate therefore means different things to different
// verifiers, and the signature covers only one of those meanings. There
// is no safe way to pick, so it is refused.
//
// Lone surrogates: a \uD800-\uDFFF escape with no partner is not a
// character. Go's decoder replaces it with U+FFFD, JavaScript keeps it,
// so the two canonicalize different strings and compute different
// digests for the same file. RFC 8785 requires rejection, and this is
// where it happens.
func validateJSONText(data []byte) error {
	if !utf8.Valid(data) {
		return fmt.Errorf("receipts: the file is not valid UTF-8")
	}
	if err := rejectLoneSurrogates(data); err != nil {
		return err
	}
	if err := rejectNumberSpellings(data); err != nil {
		return err
	}
	return rejectDuplicateMembers(data)
}

// rejectNumberSpellings refuses a number written with a fractional part
// or an exponent. Every number in this format is an integer (SPEC
// section 4), and 1, 1.0 and 1e0 are one value with three spellings: a
// typed decoder refuses two of them and a browser parser cannot tell
// them apart, so the check belongs on the text. The browser does the
// same in verifier/src/jsontext.ts.
func rejectNumberSpellings(data []byte) error {
	s := string(data)
	inString := false
	for i := 0; i < len(s); i++ {
		c := s[i]
		if inString {
			if c == '\\' {
				i++
			} else if c == '"' {
				inString = false
			}
			continue
		}
		if c == '"' {
			inString = true
			continue
		}
		if c != '-' && (c < '0' || c > '9') {
			continue
		}
		j := i
		if s[j] == '-' {
			j++
		}
		start := j
		for j < len(s) && (s[j] == '.' || s[j] == 'e' || s[j] == 'E' || s[j] == '+' || s[j] == '-' || (s[j] >= '0' && s[j] <= '9')) {
			j++
		}
		token := s[start:j]
		if strings.ContainsAny(token, ".eE") {
			return fmt.Errorf("receipts: %s is written with a fractional part or an exponent; this format carries whole numbers only", s[i:j])
		}
		i = j - 1
	}
	return nil
}

func rejectLoneSurrogates(data []byte) error {
	s := string(data)
	for i := 0; i+5 < len(s); i++ {
		if s[i] != '\\' || s[i+1] != 'u' {
			continue
		}
		// A preceding backslash means this one is escaped, not an escape.
		backslashes := 0
		for j := i - 1; j >= 0 && s[j] == '\\'; j-- {
			backslashes++
		}
		if backslashes%2 == 1 {
			continue
		}
		r, ok := parseHex4(s[i+2 : i+6])
		if !ok {
			continue // malformed escapes are the decoder's business
		}
		if !utf16.IsSurrogate(rune(r)) {
			continue
		}
		// A high surrogate is only valid immediately followed by a low one.
		if i+11 < len(s) && s[i+6] == '\\' && s[i+7] == 'u' {
			if r2, ok2 := parseHex4(s[i+8 : i+12]); ok2 {
				if utf16.DecodeRune(rune(r), rune(r2)) != 0xFFFD {
					i += 11 // consume the pair
					continue
				}
			}
		}
		return fmt.Errorf("receipts: the file contains a lone surrogate escape (\\u%04X), which is not a character", r)
	}
	return nil
}

// endValue records that a scalar value has been consumed: inside an
// object the next token is a member name again.
func endValue(stack []map[string]bool, path *[]string, expectKey *bool) {
	if len(stack) > 0 && stack[len(stack)-1] != nil {
		*expectKey = true
		if len(*path) > 0 {
			*path = (*path)[:len(*path)-1]
		}
	}
}

func parseHex4(s string) (int, bool) {
	if len(s) != 4 {
		return 0, false
	}
	n := 0
	for _, c := range s {
		n <<= 4
		switch {
		case c >= '0' && c <= '9':
			n |= int(c - '0')
		case c >= 'a' && c <= 'f':
			n |= int(c-'a') + 10
		case c >= 'A' && c <= 'F':
			n |= int(c-'A') + 10
		default:
			return 0, false
		}
	}
	return n, true
}

// rejectDuplicateMembers walks the document and refuses any object that
// names the same member twice, at any depth.
func rejectDuplicateMembers(data []byte) error {
	dec := json.NewDecoder(bytes.NewReader(data))
	dec.UseNumber()
	// One set of seen names per open object; nil marks an open array.
	var stack []map[string]bool
	var path []string
	expectKey := false

	for {
		tok, err := dec.Token()
		if err == io.EOF {
			return nil
		}
		if err != nil {
			return err // malformed JSON: the decoder reports it properly
		}
		switch t := tok.(type) {
		case json.Delim:
			switch t {
			case '{':
				stack = append(stack, map[string]bool{})
				expectKey = true
			case '[':
				stack = append(stack, nil)
				expectKey = false
			case '}', ']':
				stack = stack[:len(stack)-1]
				if len(path) > 0 {
					path = path[:len(path)-1]
				}
				expectKey = len(stack) > 0 && stack[len(stack)-1] != nil
			}
		case string:
			if expectKey && len(stack) > 0 && stack[len(stack)-1] != nil {
				seen := stack[len(stack)-1]
				if seen[t] {
					where := strings.Join(path, ".")
					if where == "" {
						where = "the top level"
					}
					return fmt.Errorf("receipts: %q appears twice in %s; a duplicate member means different things to different parsers", t, where)
				}
				seen[t] = true
				path = append(path, t)
				expectKey = false
				continue
			}
			endValue(stack, &path, &expectKey)
		default:
			endValue(stack, &path, &expectKey)
		}
	}
}
